### PR TITLE
Ticket Emails: Failed/Completed Order Template Files

### PR DESCRIPTION
**Pull request by codingmusician**
### 🎫 Ticket

[ET-1681] 
[ET-1682] 

### 🗒️ Description

This PR creates the template files for the Failed and Completed Order Emails.

### 🎥 Artifacts 

image
image

### ✔️ Checklist
- [ ] I've included a changelog entry in the readme.txt file. 
- [x] My code is tested. 
- [x] My code has [proper inline documentation](https://the-events-calendar.github.io/products-engineering/docs/code-standards/).

[ET-1682]: https://theeventscalendar.atlassian.net/browse/ET-1682?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[ET-1681]: https://theeventscalendar.atlassian.net/browse/ET-1681?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/views/v2/emails/admin-failed-order.php [hunk 1]
```diff
@@ -16,8 +16,16 @@
...
 $this->template( 'template-parts/header' );
 
-$this->template( 'template-parts/body' );
+$this->template( 'template-parts/body/title' );
+$this->template( 'template-parts/body/order/error-message' );
+$this->template( 'template-parts/body/order/purchaser-details' );
+$this->template( 'template-parts/body/order/event-title' );
+$this->template( 'template-parts/body/order/ticket-totals' );
+$this->template( 'template-parts/body/order/order-total', [ 'status' => 'failed' ] );
+$this->template( 'template-parts/body/order/payment-info' );
```
Comment (juanfra) [inline src/views/v2/emails/admin-failed-order.php:29]: The idea of the `template-parts` folder is that it will be the home for template parts that are used across templates (header, footer).

Then each template will have its own folder where we should be having the templates (atomized) for the different email templates.

For example, we'll be having folders for:

- `v2/tickets`
- `v2/rsvp`
- `v2/failed-order`
- and so forth.

The idea of the main template for each email is to have a structure of:

```
v2/template-parts/header

v2/email-template/body

v2/template-parts/footer
```

File: src/views/v2/emails/template-parts/body/order/order-total.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+ // @todo @codingmusician @juanfra Replace hardcoded data with dynamic data.
+
+?>
+<tr>
+	<td style="padding-top:20px;text-align:right" align="right">
+		<table style="display:inline-block;width:auto">
+			<tr>
+				<td style="font-size:14px;font-weight:400;line-height:24px;padding-right:10px">Order Total</td>
```
Comment (juanfra) [inline src/views/v2/emails/template-parts/body/order/order-total.php:31]: We should move as much of these inline styles to the `/v2/emails/template-parts/header/head/styles.php` template as CSS classes, so that's easier for people to customize these, instead of going through a bunch of templates. 

that way presentation is separate from content, and they can adjust whatever padding or something they like from the styles, and if they want to alter the content they go to the template itsel.

File: src/views/v2/emails/template-parts/body/order/purchaser-details.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ * @var int              $order_id              [Global] The order ID.
+ * @var bool             $is_tec_active         [Global] Whether `The Events Calendar` is active or not.
+ */
+
+// @todo @codingmusician @juanfra Replace hardcoded data with dynamic data.
+
+?>
+<tr>
+	<td>
+		<table>
```
Comment (juanfra) [inline src/views/v2/emails/template-parts/body/order/purchaser-details.php:29]: At this point, for templates in which we're having really basic layout that's supported by any email client, we should go with divs and basic HTML elements. If we end up using too many tables, the customization experience will be not ideal.

File: src/views/v2/emails/template-parts/body/order/purchaser-details.php [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+			<tr>
+				<th style="font-size:16px;font-weight:700;line-height:23px;text-align:left" align="left">
+					Order #123
+				</th>
+				<th style="font-size:16px;font-weight:700;line-height:23px;text-align:right" align="right">
+					David Hickox
+				</th>
+			</tr>
+			<tr>
+				<td style="font-size:14px;font-weight:400;line-height:23px;text-align:left" align="left">
```
Comment (juanfra) [inline src/views/v2/emails/template-parts/body/order/purchaser-details.php:39]: We also need to atomize these tiny chunks into smaller templates, that way, it's easier to customize tiny bits and maintain the whole structure. At the same time, for consistency with the other templates we have in place for the plugin.

- order/purchaser-details/order-number
- order/purchaser-details/purchaser-name
- order/purchaser-details/order-date
- order/purchaser-details/purchaser-email

**Review by juanfra — COMMENTED** [review]
Looking good, I left some comments regarding templates organization/structure/atomization, using css classes instead of too many inline styles, and maybe try to use more divs rather than having tables for everything. Having tables for everything will make it really difficult to maintain the emails coming forward.

File: src/views/v2/emails/template-parts/body/order/order-total.php (hunk 2 above)
Comment (codingmusician) [reply src/views/v2/emails/template-parts/body/order/order-total.php:31]: The reason I chose to do inline styling is that is what my research suggested when I was looing into email design. I'd much rather use stylesheets. I'll go ahead and do it for now, but we need to thoroughly test different email clients to ensure everything looks okay.

Comment (juanfra) [reply src/views/v2/emails/template-parts/body/order/order-total.php:31]: You can always check what's supported [here](https://www.caniemail.com/)

File: src/views/v2/emails/template-parts/body/order/purchaser-details.php (hunk 3 above)
Comment (juanfra) [reply src/views/v2/emails/template-parts/body/order/purchaser-details.php:29]: We can double-check if things are supported [here](https://www.caniemail.com/), for example [the `<div>` tag](https://www.caniemail.com/features/html-div/).

**Review by bordoni — DISMISSED** [review]
Can we rename `template-parts` to `components` to match usage in other parts of the code.

File: src/views/v2/emails/template-parts/body/order/purchaser-details/order-number.php [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ * @var string           $provider_id           [Global] The tickets provider class name.
+ * @var \WP_Post         $order                 [Global] The order object.
+ * @var int              $order_id              [Global] The order ID.
+ * @var bool             $is_tec_active         [Global] Whether `The Events Calendar` is active or not.
+ */
+
+// @todo @codingmusician @juanfra Replace hardcoded data with dynamic data.
+
+?>
+<th class="tec-tickets__email-table-content-order-purchaser-details-top tec-tickets__email-table-content-align-left" align="left">
```
Comment (juanfra) [inline src/views/v2/emails/template-parts/body/order/purchaser-details/order-number.php:27]: Same comment about having this as a `th`. From the design this doesn't appear to be a table header cell.

File: src/views/v2/emails/template-parts/body/order/ticket-totals/header-row.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ * @var bool             $is_tec_active         [Global] Whether `The Events Calendar` is active or not.
+ */
+
+ // @todo @codingmusician @juanfra Replace hardcoded data with dynamic data.
+
+?>
+<tr class="tec-tickets__email-table-content-order-ticket-totals-header-row">
+	<td style="width: 80%" class="tec-tickets__email-table-content-order-ticket-totals-cell tec-tickets__email-table-content-order-align-left" align="left">Ticket</td>
+	<td class="tec-tickets__email-table-content-order-ticket-totals-cell tec-tickets__email-table-content-order-align-center" align="center">Qty</td>
+	<td class="tec-tickets__email-table-content-order-ticket-totals-cell tec-tickets__email-table-content-order-align-right" align="right">Price</td>
```
Comment (juanfra) [inline src/views/v2/emails/template-parts/body/order/ticket-totals/header-row.php:30]: These strings will not come from the dynamic data, we should use gettext functions for them.

**Review by juanfra — COMMENTED** [review]
Looking good! I guess we'll continue working on these and maybe modifying some structure once we "dynamize" them 😉

File: src/views/v2/emails/template-parts/body/order/attendees-table.php [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+ * @var Tribe_Template  $this  Current template object.
+ * @var Module           $provider              [Global] The tickets provider instance.
+ * @var string           $provider_id           [Global] The tickets provider class name.
+ * @var \WP_Post         $order                 [Global] The order object.
+ * @var int              $order_id              [Global] The order ID.
+ * @var bool             $is_tec_active         [Global] Whether `The Events Calendar` is active or not.
+ */
+
+// @todo @codingmusician @juanfra Replace hardcoded data with dynamic data.
+$attendees = [
```
[the remark this answers is not in the document]
Comment (codingmusician) [reply src/views/v2/emails/template-parts/body/order/attendees-table.php]: I went ahead and `dynamized` all the data.

File: src/Tickets/Emails/Email/Completed_Order.php [hunk 8]
```diff
@@ -7,6 +7,8 @@
 
 namespace TEC\Tickets\Emails\Email;
 
+use Automattic\WooCommerce\StoreApi\Formatters\CurrencyFormatter;
+use TEC\Tickets\Commerce\Utils\Currency;
```
Comment (juanfra) [inline src/Tickets/Emails/Email/Completed_Order.php:11]: Are we using any of these?

File: src/Tickets/Emails/Email_Abstract.php [hunk 9]
```diff
@@ -434,8 +434,8 @@ public function get_placeholders(): array {
 	 */
 	public function format_string( $string ): string {
 		$placeholders = $this->get_placeholders();
-		$find         = array_keys( $placeholders );
-		$replace      = array_values( $placeholders );
+		$find         = array_keys( $this->placeholders );
+		$replace      = array_values( $this->placeholders );
```
Comment (juanfra) [inline src/Tickets/Emails/Email_Abstract.php:438]: The point of getting the placeholders from `get_placeholders()`, is that it is returning the filtered placeholders, so we can hook from other places. [Such as TEC](https://github.com/the-events-calendar/the-events-calendar/pull/4196/files#diff-7a9e4d535269e6b251e2a56a46c6e38ae4d0bcb385b87def7227e79e43664337R33). We should keep it like that and revert this change.

File: src/Tickets/Emails/Email_Abstract.php [hunk 10]
```diff
@@ -828,4 +828,101 @@ public function get_post_type_data(): array {
...
+			'provider'   => 'Stripe',
+			'status'     => 'success',
+			'post_title' => 'Black Midi with Special Guests Chat Pile and Apprehend',
+			'total'      => '$100.00',
+			'purchaser'  => [
+				'name' => 'John Doe',
+				'email' => 'john@doe.com',
+			],
+			'created' => 'March 1, 2023',
+			'id'      => 123,
```
Comment (juanfra) [inline src/Tickets/Emails/Email_Abstract.php:850]: The strings that are for presentation purposes need to be translatable (Dates, post title, purchaser).

File: src/Tickets/Emails/Email_Abstract.php [hunk 11]
```diff
@@ -828,4 +828,101 @@ public function get_post_type_data(): array {
...
 	}
+
+	/**
+	 * Get a successful order array for preview.
+	 * 
+	 * @since TBD
+	 *
+	 * @return array
+	 */
+	public function get_preview_order_successful(): array {
```
Comment (juanfra) [inline src/Tickets/Emails/Email_Abstract.php:839]: Also, this shouldn't live in abstract. The goal of an abstract class is to be extended, storing common functionality that's applicable to any class that's extending it. If we're using `order` data in one or two emails, then these things should live there. Not in the abstract class.

File: src/Tickets/Emails/Email_Abstract.php [hunk 12]
```diff
@@ -828,4 +828,101 @@ public function get_post_type_data(): array {
...
+	}
+
+	/**
+	 * Get a failed order array for preview.
+	 * 
+	 * @since TBD
+	 *
+	 * @return array
+	 */
+	public function get_preview_order_failed(): array {
```
Comment (juanfra) [inline src/Tickets/Emails/Email_Abstract.php:861]: Anything order related should be outside of the abstract class.

File: src/views/v2/emails/template-parts/body/order/attendees-table/attendee-info.php [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+if ( empty( $attendee ) ) {
+	return;
+}
+
+?>
+<tr class="tec-tickets__email-table-content-order-attendee-info-row">
+	<td class="tec-tickets__email-table-content-order-attendee-info tec-tickets__email-table-content-align-left" align="left">
+		<div><?php esc_html_e( $attendee['name'] ); ?></div>
+		<div><?php esc_html_e( $attendee['email'] ); ?></div>
```
Comment (juanfra) [inline src/views/v2/emails/template-parts/body/order/attendees-table/attendee-info.php:32]: Should we make a safe check to see if they keys exist in case there's a particular case where these don't exist? (it'll break up the email functionality.

File: src/views/v2/emails/template-parts/body/order/ticket-totals.php [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * @var Tribe_Template  $this  Current template object.
+ * @var Module           $provider              [Global] The tickets provider instance.
+ * @var string           $provider_id           [Global] The tickets provider class name.
+ * @var array            $order                 [Global] The order object.
+ * @var int              $order_id              [Global] The order ID.
+ * @var bool             $is_tec_active         [Global] Whether `The Events Calendar` is active or not.
+ */
+
+// @todo @codingmusician @juanfra Replace hardcoded data with dynamic data.
```
Comment (juanfra) [inline src/views/v2/emails/template-parts/body/order/ticket-totals.php:24]:
```suggestion
```

File: src/Tickets/Emails/Email_Abstract.php (hunk 9 above)
Comment (codingmusician) [reply src/Tickets/Emails/Email_Abstract.php:438]: Ah. I was getting an error. Looking back at it now, I see it was because of the `@return` type for the `get_placeholders` method. Sorry about that. I'll fix it here.

File: src/Tickets/Emails/Email/Completed_Order.php (hunk 8 above)
Comment (codingmusician) [reply src/Tickets/Emails/Email/Completed_Order.php]: I was toying around with it, but decided not to use them. Forgot to remove those lines, though. Good catch.

File: src/Tickets/Emails/Email_Abstract.php (hunk 10 above)
Comment (codingmusician) [reply src/Tickets/Emails/Email_Abstract.php:850]: I wasn't quite sure that this would turn out to be the final format of the arrays, so I didn't go as far as to make things translatable. I'll do that, though.

File: src/views/v2/emails/template-parts/body/order/attendees-table/attendee-info.php (hunk 13 above)
Comment (codingmusician) [reply src/views/v2/emails/template-parts/body/order/attendees-table/attendee-info.php]: Good call. I actually split each of them out into their own template file and bailed if they were empty.

File: src/Tickets/Emails/Email_Abstract.php (hunk 12 above)
Comment (codingmusician) [reply src/Tickets/Emails/Email_Abstract.php]: I moved all preview data into its own class called `Preview_Data`. Any of the emails can fetch the data from there.

File: src/Tickets/Emails/Email_Abstract.php (hunk 11 above)
Comment (codingmusician) [reply src/Tickets/Emails/Email_Abstract.php]: I moved all preview data into its own class called `Preview_Data`. Any of the emails can fetch the data from there.

File: src/Tickets/Emails/Preview_Data.php [hunk 15]
```diff
@@ -0,0 +1,106 @@
+<?php
+/**
+ * Tickets Emails Preview Data class.
+ *
+ * @since TBD
+ *
+ * @package TEC\Tickets\Emails
+ */
+
+namespace TEC\Tickets\Emails;
```
Comment (juanfra) [inline src/Tickets/Emails/Preview_Data.php:10]: Given that we have the `Preview_Modal` in the `Admin` folder, would it also make more sense to have the `Preview_Data` there?

File: src/Tickets/Emails/Preview_Data.php [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+	/**
+	 * Get Order preview data.
+	 * 
+	 * @since TBD
+	 *
+	 * @param string $status Status of order.
+	 * 
+	 * @return array
+	 */
+	static function get_order( $status = 'success' ) {
```
Comment (juanfra) [inline src/Tickets/Emails/Preview_Data.php:30]: We must declare visibility. And return type. I'd follow the pattern of receiving an array as parameter, and then looking for the `success` key.

File: src/Tickets/Emails/Preview_Data.php [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+	}
+
+	/**
+	 * Get Attendees preview data.
+	 * 
+	 * @since TBD
+	 * 
+	 * @return array
+	 */
+	static function get_attendees() {
```
Comment (juanfra) [inline src/Tickets/Emails/Preview_Data.php:52]: We have to declare visibility for methods. We are now also including the return type when possible.

File: src/Tickets/Emails/Preview_Data.php [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+	}
+
+	/**
+	 * Get Tickets preview data.
+	 * 
+	 * @since TBD
+	 * 
+	 * @return array
+	 */
+	static function get_tickets() {
```
Comment (juanfra) [inline src/Tickets/Emails/Preview_Data.php:96]: We have to declare visibility for methods. We are now also including the return type when possible.

File: src/views/v2/emails/template-parts/body/order/purchaser-details/order-number.php [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ * @var Tribe_Template  $this  Current template object.
+ * @var Module           $provider              [Global] The tickets provider instance.
+ * @var string           $provider_id           [Global] The tickets provider class name.
+ * @var array            $order                 [Global] The order object.
+ * @var int              $order_id              [Global] The order ID.
+ * @var bool             $is_tec_active         [Global] Whether `The Events Calendar` is active or not.
+ */
+
+$order_number = empty( $order['id'] ) ? 0 : $order['id'];
```
Comment (juanfra) [inline src/views/v2/emails/template-parts/body/order/purchaser-details/order-number.php:24]: Should we cast this as integer so we're safe on using it as `%d` on the `sprintf`?

File: src/Tickets/Emails/Preview_Data.php (hunk 15 above)
Comment (codingmusician) [reply src/Tickets/Emails/Preview_Data.php]: Done

File: src/Tickets/Emails/Preview_Data.php (hunk 16 above)
Comment (codingmusician) [reply src/Tickets/Emails/Preview_Data.php]: I changed all the methods to where it accepts an `$args` parameter so that you can override any of the array items. Good call!

File: src/Tickets/Emails/Preview_Data.php (hunk 17 above)
Comment (codingmusician) [reply src/Tickets/Emails/Preview_Data.php]: Done

File: src/Tickets/Emails/Preview_Data.php (hunk 18 above)
Comment (codingmusician) [reply src/Tickets/Emails/Preview_Data.php]: Done

File: src/views/v2/emails/template-parts/body/order/purchaser-details/order-number.php (hunk 19 above)
Comment (codingmusician) [reply src/views/v2/emails/template-parts/body/order/purchaser-details/order-number.php]: Done

File: src/views/v2/emails/admin-failed-order.php [hunk 20]
```diff
@@ -1,23 +1,22 @@
 <?php
 /**
- * Event Tickets Emails: Failed Order Template.
+ * Event Tickets Emails: Failed Order Template
  *
  * Override this template in your own theme by creating a file at:
- * [your-theme]/tribe/tickets/v2/emails/admin-failed-order.php
+ * [your-theme]/tribe/tickets/v2/emails/failed-order.php
```
Comment (juanfra) [inline src/views/v2/emails/admin-failed-order.php:6]:
```suggestion
 * [your-theme]/tribe/tickets/v2/emails/admin-failed-order.php
```

File: src/views/v2/emails/admin-new-order.php [hunk 21]
```diff
@@ -1,23 +1,22 @@
 <?php
 /**
- * Event Tickets Emails: New Order Template.
+ * Event Tickets Emails: New Order Template
  *
  * Override this template in your own theme by creating a file at:
- * [your-theme]/tribe/tickets/v2/emails/admin-new-order.php
+ * [your-theme]/tribe/tickets/v2/emails/new-order.php
```
Comment (juanfra) [inline src/views/v2/emails/admin-new-order.php:6]:
```suggestion
 * [your-theme]/tribe/tickets/v2/emails/admin-new-order.php
```

File: src/views/v2/emails/template-parts/body/order/purchaser-details.php [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * @var Tribe_Template  $this  Current template object.
+ * @var Module           $provider              [Global] The tickets provider instance.
+ * @var string           $provider_id           [Global] The tickets provider class name.
+ * @var array            $order                 [Global] The order object.
+ * @var int              $order_id              [Global] The order ID.
+ * @var bool             $is_tec_active         [Global] Whether `The Events Calendar` is active or not.
+ */
+
+if( empty( $order ) ) {
```
Comment (juanfra) [inline src/views/v2/emails/template-parts/body/order/purchaser-details.php:24]:
```suggestion
if ( empty( $order ) ) {
```

**Review by juanfra — APPROVED** [review]
